### PR TITLE
Create routes for user creation and login 

### PR DESCRIPTION
**Pull request by DarkPitoune**
[This PR is a bit urgent as it will unlock development on the front end]
## Purpose

To create a user and to login as a user

## Proposal
There is one view for each route. The auth takes as much of the native features of django as possible

### CODE REVIEW

File: sandbox/settings.py [hunk 1]
```diff
@@ -95,6 +95,24 @@ class Base(DRFMixin, MagnifyCoreConfigurationMixin, Configuration):
...
     AUTH_USER_MODEL = "core.User"
 
+    AUTH_PASSWORD_VALIDATORS = [
+        {
+            "NAME": "django.contrib.auth.password_validation.UserAttributeSimilarityValidator",
+        },
+        {
+            "NAME": "django.contrib.auth.password_validation.MinimumLengthValidator",
+            "OPTIONS": {
+                "min_length": 8,
```
Comment (lunika) [inline sandbox/settings.py:105]: Add the ability to override this value with an environment variable using `values.Value`

File: src/magnify/apps/core/models.py [hunk 2]
```diff
@@ -8,7 +8,48 @@
...
+        "^[a-zA-Z0-9_-]{3,16}$",
+        message="""Username must be between 3 and 16 characters long
+            and contain only letters, numbers, underscores and hyphens.""",
+    )
+
+    username = models.CharField(
+        _("username"),
+        max_length=150,
+        unique=True,
+        validators=[username_validator],
```
Comment (lunika) [inline src/magnify/apps/core/models.py:23]: I'm not fan of having validation on the model. In an API I prefer having them on the serializer unless you plan to use this model in a management command.

File: src/magnify/apps/core/models.py [hunk 3]
```diff
@@ -8,7 +8,48 @@
...
+            RegexValidator(
+                regex=r"^[a-zA-Z '-]+$",
+                message=_("Only letters, spaces, apostrophe and hyphen are allowed"),
+            )
+        ],
+    )
+    email = models.EmailField(
+        max_length=255,
+        unique=True,
+        validators=[
```
Comment (lunika) [inline src/magnify/apps/core/models.py:43]: IMO this is useless and already managed by DRF and django https://www.django-rest-framework.org/api-guide/fields/#emailfield

File: src/magnify/apps/core/serializers/users.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
+"""Serializers for users views"""
+
+from rest_framework import serializers
+
+
+class UserCreateSerializer(serializers.Serializer):
```
Comment (lunika) [inline src/magnify/apps/core/serializers/users.py:6]: Why not using a ModelSerializer mapping it with the user model ? Doing this you have only one serializer https://www.django-rest-framework.org/api-guide/serializers/#modelserializer

File: src/magnify/apps/core/views/user.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+from magnify.apps.core.models import User
+from magnify.apps.core.serializers import (
+    UserCreateResponseErrorSerializer,
+    UserCreateResponseSuccessSerializer,
+    UserCreateSerializer,
+)
+
+
+class UserView(APIView):
```
Comment (lunika) [inline src/magnify/apps/core/views/user.py:19]: Consider using a ModelViewset to not implement everything https://www.django-rest-framework.org/api-guide/viewsets/#modelviewset

**Comment by lunika** [comment]
Lot of code in this PR can be removed using DRF views and serializer mapped on the model.

File: src/magnify/apps/core/models.py (hunk 2 above)
Comment (sampaccoud) [reply src/magnify/apps/core/models.py:23]: why are you against model validation for this case? Shouldn't it even be a database constraint if we really want to be sure our users are properly named?

Comment (lunika) [reply src/magnify/apps/core/models.py:23]: Most of time validators on the database avoid future modification. Once modified, part of your data will become invalid and you will not be able to modify the all model without doing a sql migration.

Comment (sampaccoud) [reply src/magnify/apps/core/models.py:23]: I would consider this a feature but we can discuss irl.

**Comment by DarkPitoune** [comment]
@sampaccoud I added jwt auth on some routes so the test now fail. They are not required but it would be nice if it was patched before merge

File: sandbox/settings.py [hunk 6]
```diff
@@ -208,6 +235,7 @@ class Base(DRFMixin, MagnifyCoreConfigurationMixin, Configuration):
         "dockerflow.django",
         "parler",
         "rest_framework",
+        "rest_framework.authtoken",
```
Comment (sampaccoud) [inline sandbox/settings.py:238]: no more needed

File: src/magnify/apps/core/urls.py [hunk 7]
```diff
@@ -13,15 +16,25 @@
         title="Magnify API",
         default_version="v1",
         description="""This is the schema for the Magnify API.
-            This app is used in the sandbox folder of the project.""",
+
+            All routes are protected by the JWT authentication except
+            for "login", "refresh" and "create user".""",
     ),
     public=True,
+    url=os.environ.get("BACK_URL"),
```
Comment (sampaccoud) [inline src/magnify/apps/core/urls.py:24]: you should go through a Django setting (see existing settings with value.Value())

File: src/magnify/apps/core/views/room_token_view.py [hunk 8]
```diff
@@ -53,8 +54,11 @@ class RoomTokenView(APIView):
     """View for token fetching based on a guest user and a specific room
     This redirects to jitsi with the token"""
 
+    permission_classes = (IsAuthenticated,)
```
Comment (sampaccoud) [inline src/magnify/apps/core/views/room_token_view.py:57]: This is why your tests are failing. You should not force authentication, you should test in the view that if the user is connected and deduct from this what s.he can do or not.

File: src/magnify/apps/core/models.py (hunk 2 above)
Comment (DarkPitoune) [reply src/magnify/apps/core/models.py:23]: I really don't know how to add validators in the serializer. If the feature is seen as necessary, some help would be welcome ;)

File: src/magnify/apps/core/serializers/users.py (hunk 4 above)
Comment (DarkPitoune) [reply src/magnify/apps/core/serializers/users.py:6]: Tks a lot for this review, I'm working on it
I still have some issues with the response from the API. In order to have good documentation, I'm creating a serializer for each possible response, which is very heavy. Do you know if there is a better way to handle this or should I just not care ?

File: sandbox/settings.py (hunk 1 above)
Comment (DarkPitoune) [reply sandbox/settings.py]: A value does not work for it triggers an error when django tries to compare an integer and an intergerValue. It needs an integer

**Comment by DarkPitoune** [comment]
All reviews were implemented. Everything is tested by hand but avatar. I'm writing tests to ensure everything but I'm ready for reviews on the code :)

**Comment by DarkPitoune** [comment]
PR is fully completed and tested. I don't plan on writing the avatar endpoints and tests on this PR
Open for reviews

File: CHANGELOG.md [hunk 9]
```diff
@@ -10,6 +10,7 @@ Versioning](https://semver.org/spec/v2.0.0.html).
 
 - Add models to handle rooms, meetings, labels and groups
 - Add a swagger view to help with front end development
+- Add a route for user creation and login
```
Comment (kernicPanel) [inline CHANGELOG.md:13]:
```suggestion
- Add routes for user creation, modification and login
```

**Review by kernicPanel — APPROVED** [review]

**Comment by DarkPitoune** [comment]
PR is unusable as is, another one is to be opened soon

**Pull request closed without merging**